akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/PersistentFsmToTypedMigrationSpec.scala

```diff
@@ -0,0 +1,253 @@
+import akka.persistence.typed.{ EventAdapter, EventSeq, ExpectingReply, PersistenceId, SnapshotAdapter }
+import akka.persistence.typed.scaladsl.Effect
+import akka.persistence.typed.scaladsl.EventSourcedBehavior
+import akka.persistence.typed.scaladsl.PersistentFSMMigration
+import com.typesafe.config.ConfigFactory
+import org.scalatest.WordSpec
+import org.scalatest.concurrent.ScalaFutures
+
+object PersistentFsmToTypedMigrationSpec {
+  val config = ConfigFactory.parseString(s"""
+    akka.persistence.journal.leveldb.dir = "target/typed-persistence-${UUID.randomUUID().toString}"
+    akka.persistence.journal.plugin = "akka.persistence.journal.leveldb"
+    akka.persistence.snapshot-store.plugin = "akka.persistence.snapshot-store.local"
+    akka.persistence.snapshot-store.local.dir = "target/typed-persistence-${UUID.randomUUID().toString}"
+    """)
+
+}
+
+object ShoppingCartActor {
+
+  def apply(pid: PersistenceId) = behavior(pid)
+
+  //#commands
+  sealed trait Command
+  case class AddItem(item: Item) extends Command
+  case object Buy extends Command
+  case object Leave extends Command
+  case class GetCurrentCart(replyTo: ActorRef[ShoppingCart]) extends Command with ExpectingReply[ShoppingCart]
+  //#commands
+
+  //#state
+  sealed trait State
+  case class LookingAround(cart: ShoppingCart) extends State
+  case class Shopping(cart: ShoppingCart) extends State
+  case class Inactive(cart: ShoppingCart) extends State
+  case class Paid(cart: ShoppingCart) extends State
+  //#state
+
+  //#snapshot-adapter
+  val persistentFSMSnapshotAdapter: SnapshotAdapter[State] = PersistentFSMMigration.snapshotAdapter[State] {
+    case (stateIdentifier, data, _) =>
+      val cart = data.asInstanceOf[ShoppingCart]
+      stateIdentifier match {
+        case "Looking Around" => LookingAround(cart)
+        case "Shopping"       => Shopping(cart)
+        case "Inactive"       => Inactive(cart)
+        case "Paid"           => Paid(cart)
+        case id               => throw new IllegalStateException(s"Unexpected state identifier $id")
+      }
+  }
+  //#snapshot-adapter
+
+  //#event-adapter
+  class PersistentFsmEventAdapter extends EventAdapter[DomainEvent, Any] {
+    override def toJournal(e: DomainEvent): Any = e
+    override def manifest(event: DomainEvent): String = ""
+    override def fromJournal(journalEvent: Any, manifest: String): EventSeq[DomainEvent] = {
+      journalEvent match {
+        case _: StateChangeEvent =>
+          // Alternatively this could be converted into a private event if the state
+          // information is required as it can't be inferred from the events
+          EventSeq.empty
+        case other =>
+          // If using a new domain event model the conversion would happen here
+          EventSeq.single(other.asInstanceOf[DomainEvent])
+      }
+
+    }
+  }
+  //#event-adapter
+
+  //#command-handler
+  val commandHandler: (State, Command) => Effect[DomainEvent, State] = (state, command) => {
```

Comment:
might be easier with a method (that is lifted to function)?
```
def commandHandler(state: State, command: Command): Effect[DomainEvent, State] = {
```

Reply:
Yep, now i've added timers i'll make this a curried